File: AI/AI Development/modules/00e-go-platform/practice/middleware.go
package main

import (
	"context"
	"net/http"

	"github.com/google/uuid"
)

type contextKey string

const tenantIDKey contextKey = "tenant_id"

// TODO A2: generate UUID per request, set response header X-Request-ID
func requestIDMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		requestId:=r.Header.Get("X-Request-ID")
		if(requestId==""){
			requestId=uuid.New().String()
		}
		w.Header().Set("X-Request-ID",requestId)
		next.ServeHTTP(w, r)
	})
}

// TODO A2: read X-API-Key header; if missing/invalid → 401; else set tenant_id in context
func tenantAuthMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		next.ServeHTTP(w, r)
	})
}

func tenantFromContext(ctx context.Context) string {
	// TODO: read tenant_id from context after middleware sets it
	return ""
}

File: AI/AI Development/modules/00c-fastapi/practice/app/middleware.py
"""A3 — Request ID middleware."""

import uuid

# TODO: import BaseHTTPMiddleware from starlette.middleware.base
from starlette.middleware.base import BaseHTTPMiddleware


class RequestIDMiddleware(BaseHTTPMiddleware):
    """Add X-Request-ID to every response."""

    # TODO: subclass BaseHTTPMiddleware
    # TODO: in dispatch: read X-Request-ID from request or generate uuid4
    async def dispatch(self, request, call_next):
        request_id = request.headers.get("X-Request-ID") or str(uuid.uuid4())
    # TODO: await call_next(request), set response header X-Request-ID
        response= await call_next(request);
        response.headers["X-Request-ID"]=request_id
        return response
    pass
